### PR TITLE
Visualize reworked v2

### PR DESCRIPTION
**Pull request by NormannK**
new structured class to visualize the results in a pdf, dividing plot generation from data ingestion.
see example usage in the class file.

- usage of templates for each graph type
- clean, automated page layout

see #178 for example files

Todo:

- [x] fix 2 tests
- [x] add the Financial Overview graph
- [x] generate the graphs for the test in the test folder
- [x] fix all the mypy stuff
- [x] remove the old visualize
- [x] fix the tests for that

### CODE REVIEW

File: src/akkudoktoreos/utils/visualize.py [hunk 1]
```diff
@@ -0,0 +1,489 @@
+import os
+from collections.abc import Sequence
+from typing import Callable, List, Optional, Union
```
Comment (Lasall) [inline src/akkudoktoreos/utils/visualize.py:3]: typing.List -> list

**Comment by Lasall** [comment]
I think we should put some version information into the PDF (defaults to git sha)

File: src/akkudoktoreos/utils/visualize.py [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+from typing import Callable, Optional, Union
+
+import matplotlib.pyplot as plt
+import numpy as np
+from matplotlib.backends.backend_pdf import PdfPages
+
+from akkudoktoreos.config.config import get_config
+
+
+class VisualizationReport:
```
Comment (Lasall) [inline src/akkudoktoreos/utils/visualize.py:12]: inherit ConfigMixin and then just use self.config (and drop get_config)

File: src/akkudoktoreos/utils/visualize.py [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+        ],
+        markers=["o", "o", "x", "^", "^"],
+        line_styles=["-", "--", ":", "-.", "-"],
+    )
+    report.finalize_group()
+
+    # Group 3:
+    report.create_line_chart(
+        None,
+        [results["result"]["akku_soc_pro_stunde"], results["result"]["EAuto_SoC_pro_Stunde"]],
```
Comment (Lasall) [inline src/akkudoktoreos/utils/visualize.py:390]: add legend (labels?) for those. Can eauto_soc_pro_stunde be none?

File: src/akkudoktoreos/utils/visualize.py [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+        data, labels=["Group 1"], title="Violin Plot", xlabel="Group", ylabel="Values"
+    )
+
+    report.finalize_group()  # Finalize the third group of charts
+
+    # Generate the PDF report
+    report.generate_pdf()
+
+
+from akkudoktoreos.optimization.genetic import OptimizationParameters  # circular import
```
Comment (Lasall) [inline src/akkudoktoreos/utils/visualize.py:317]: move to the top

File: src/akkudoktoreos/utils/visualize.py [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+        """Generate the PDF report with all the added chart groups."""
+        self._initialize_pdf()  # Initialize the PDF
+
+        for group in self.groups:
+            self._save_group_to_pdf(group)  # Save each group to the PDF
+
+        self.pdf_pages.close()  # Close the PDF to finalize the report
+
+
+if __name__ == "__main__":
```
Comment (Lasall) [inline src/akkudoktoreos/utils/visualize.py:240]: move to the very bottom

File: tests/test_visualize.py [hunk 6]
```diff
@@ -1,36 +1,36 @@
...
+
+
+def test_generate_pdf_main():
+    # Delete the old generated file if it exists
+    if os.path.isfile(output_file):
+        os.remove(output_file)
+
+    # Execute the __main__ block of visualize.py by running it as a script
+    script_path = Path(__file__).parent.parent / "src" / "akkudoktoreos" / "utils" / "visualize.py"
+    subprocess.run(["python", str(script_path)], check=True)
```
Comment (Lasall) [inline tests/test_visualize.py:27]: should test the generation by function call (then also all the pytest.fixtures will work, e.g. for temporary config etc.).

File: src/akkudoktoreos/utils/visualize.py (hunk 5 above)
Comment (NormannK) [reply src/akkudoktoreos/utils/visualize.py]: done

File: src/akkudoktoreos/utils/visualize.py (hunk 4 above)
Comment (NormannK) [reply src/akkudoktoreos/utils/visualize.py]: done

File: src/akkudoktoreos/utils/visualize.py (hunk 3 above)
Comment (NormannK) [reply src/akkudoktoreos/utils/visualize.py:311]: added

File: src/akkudoktoreos/utils/visualize.py (hunk 2 above)
Comment (NormannK) [reply src/akkudoktoreos/utils/visualize.py]: done

**Review by Lasall — APPROVED** [review]

**Pull request merged**